### ISSUE TITLE
Investigate SM103 launch geometry for 4-bit dequantization

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/bitsandbytes#5]
## Problem

The CUDA NF4/FP4 blockwise dequantization path currently uses one fixed launch geometry for every architecture and output dtype. In `csrc/ops.cu`, each launch instantiates `kDequantizeBlockwise<T, 512, 64, 8, ...>` with 64 threads and emits up to 1,024 dequantized 4-bit values per CTA. That geometry predates the current branchless decode implementation and has not been calibrated on SM103/B300.

A larger tile or CTA could reduce CTA scheduling and CUB/barrier overhead for large weight tensors, but it could also increase register/shared-memory pressure or hurt occupancy. This issue treats an SM103-specific launch improvement as a measurement-gated hypothesis, not an assumed win.

Standalone dequantization is relevant beyond model loading: `MatMul4Bit.backward` dequantizes the weight for activation gradients, and existing fallback/deprecated-layout paths dequantize before `torch.nn.functional.linear`. The investigation is independent of fork issues #1 and #3, which concern two different GEMM dispatch layers.

## Scope

- Baseline: `bitsandbytes-foundation/bitsandbytes@95f9af309d4d5793847169c39288dcd3fcbdf564`.
- Measure the current 64-thread/512-packed-byte launch against a small set of plausible larger CTA/tile variants built from the same baseline and run in the same B300 allocation.
- Keep forced variants in private benchmark/build scaffolding. Do not expose a forcing API.
- Only if measurements identify a coherent winning region, select the narrowest exact-SM103 launch rule for NF4/FP4 dequantization. Any device-property lookup on the hot path must use an existing cached mechanism or equivalent one-time state rather than an uncached query per call.
- Preserve the current decode arithmetic, packed format, Python/public API, and behavior on every non-SM103 CUDA device and on HIP/XPU/CPU backends.

The implementation may parameterize/add only the template instantiations needed by the measured winner. It must not redesign blockwise dequantization or change General8bit behavior.

## Correctness oracle

Compare every forced variant and final automatic dispatch with the current upstream kernel and the repository's existing dequantization/reference checks.

- Cover FP16, BF16, and FP32 outputs; NF4 and FP4; blocksize 64 as the primary case plus representative 32/128/256 and larger-block controls.
- Include empty/small inputs, realistic weight shapes, odd element counts, and lengths immediately below/at/above candidate tile boundaries.
- Verify shape, dtype, finite outputs, `.out` behavior, storage/non-contiguous views exercised by current tests, and repository `torch.testing.assert_close` tolerances. Require exact equality when changing only launch geometry demonstrably preserves it; do not invent stricter numerical promises without evidence.
- Run focused CUDA op/functional dequantization tests, affected 4-bit autograd tests, a CUDA build, and the full pre-commit suite. GPU compilation and tests run only through Slurm.

## B300 benchmark and acceptance

Use one B300/SM103 allocation and record the exact commit, node/GPU, driver, CUDA compiler, PyTorch/bitsandbytes versions, commands, target architecture, and raw logs/results. Build baseline and forced variants from isolated source trees with the same compiler and flags.

- Measure direct NF4/FP4 dequantization across latency-scale inputs, tile-boundary tails, and representative weight shapes such as 4096x4096, 4096x11008 and its reverse, 5120x13824 and its reverse, 4096x14336 and its reverse, 7168/8192 square, plus a large vocabulary projection if memory permits.
- Use FP16/BF16/FP32 output, with blocksize 64 primary and adjacent blocksize controls.
- Interleave variants with at least 20 warmups and 7 repeated batches of enough CUDA-event samples for stable medians. Report median and p10/p90, complete/round-level samples, and effective bandwidth rather than a best run.
- Also measure at least one recurrent `MatMul4Bit` backward case and a dequantize-plus-linear control without changing their dispatch, so a direct-kernel win is tied to user-visible work.

Proceed only if one geometry wins by more than run-to-run noise across a coherent set of realistic sizes and both NF4/FP4, produces a material improvement in a recurrent or end-to-end control, preserves correctness, and avoids meaningful regressions in adjacent/tail cells. Use the repository's 5% decision band as the default materiality threshold. If gains are isolated, disappear end to end, or require a broad architecture/general dispatch framework, record a no-go instead of adding a heuristic.

## Non-goals

No decode-math or quantization-format change; no Python fused-vs-fallback policy change; no 4-bit GEMM SIMT/MMA change; no new public API; no General8bit tuning; no extrapolation to SM100, other CUDA architectures, or non-CUDA backends without their own evidence; no dependency on the unmerged cycle-1 or cycle-2 branches.

**Comment by heiheiha798** [comment]
## B300 result: NO-GO

The launch-geometry hypothesis did not meet the issue's 5% performance gate, so I am not proposing a source change or draft PR.

### Environment and method

- Baseline commit: `95f9af309d4d5793847169c39288dcd3fcbdf564` (clean worktree, independently based on upstream `main`)
- Slurm job: `4869` on `dev-slurm`
- GPU: NVIDIA B300 SXM6 AC, CC 10.3, 148 SMs
- PyTorch: `2.13.0+cu130`; CUDA toolkit/NVCC: `13.0.88`; driver: `580.126.09`
- Build: native `COMPUTE_CAPABILITY=103`, Release/Ninja, identical flags and isolated source/build trees for every variant
- Measurement: variants interleaved in one allocation, 20 warmups per round, 7 rounds x 100 CUDA-event samples per variant; JSONL retains all samples, round medians, median/p10/p90, and effective bandwidth

The forced private variants preserved the same decode math and changed only the packed-byte tile, CTA threads, and items per thread:

| Variant | Packed bytes/CTA | Threads | Bytes/thread |
| --- | ---: | ---: | ---: |
| baseline | 512 | 64 | 8 |
| candidate A | 512 | 128 | 4 |
| candidate B | 1024 | 128 | 8 |
| candidate C | 1024 | 256 | 4 |
| candidate D | 2048 | 256 | 8 |

Exact benchmark invocation:

```bash
python benchmark-evidence/cycle3/sm103_dequant_geometry.py \
  --output benchmark-evidence/cycle3/sm103-dequant-geometry-job-4869.jsonl \
  --build-root /tmp/bnb-sm103-dequant-4869/benchmark-build \
  --compute-capability 103 \
  --warmup 20 \
  --repetitions 100 \
  --rounds 7
```

Raw workspace evidence:

- `benchmark-evidence/cycle3/slurm-4869-dequant-geometry.out`
- `benchmark-evidence/cycle3/sm103-dequant-geometry-job-4869.jsonl`

Job `4868` is not performance evidence. Its five builds completed, but the private harness failed before the first latency sample because it used the wrong ctypes dtype suffix. Job `4869` contains the corrected complete run.

### Direct dequantization

The run covered 156 cells:

- 54 realistic blocksize-64 cells: nine matrix shapes, FP16/BF16/FP32 output, NF4/FP4
- 78 latency/tail cells: odd counts and values immediately around 64, 1024, 2048, and 4096 output elements
- 24 adjacent/larger-block controls: blocksize 32/128/256/4096 on 4096x4096

No candidate was more than 5% faster in any direct cell. Across the 54 realistic cells, no candidate won by more than 5% in any of the 378 per-round comparisons either. Ratios below are `baseline median / candidate median`, so values below 1.0 are regressions:

| Candidate | Realistic median ratio | Range | Cells >5% faster | Characterization |
| --- | ---: | ---: | ---: | --- |
| 1024 packed / 128 threads | 0.99958x | 0.95880-1.00158x | 0/54 | neutral overall; up to 4.1% slower |
| 512 packed / 128 threads | 0.94859x | 0.90831-1.00000x | 0/54 | recurrent regression |
| 1024 packed / 256 threads | 0.91319x | 0.88287-0.98865x | 0/54 | broad regression |
| 2048 packed / 256 threads | 0.96140x | 0.90222-1.00158x | 0/54 | mostly slower |

Representative baseline medians were 0.02029 ms / 2119 GB/s for 4096x4096 FP16 NF4, 0.05715 ms / 2633 GB/s for 4096x14336 FP16 NF4, and 0.06506 ms / 2643 GB/s for 8192x8192 FP16 NF4. The neutral 1024/128 candidate was respectively 0.02026, 0.05715, and 0.06522 ms, which is not a material improvement.

### Recurrent and end-to-end controls

For both 4096x14336 orientations, FP16/BF16, and NF4/FP4:

- 8 preallocated dequantize-plus-matmul cells: no candidate won by more than 5%.
- 8 actual `MatMul4Bit.backward` cells: no candidate won by more than 5%.
- The neutral 1024/128 candidate ranged from 0.97738-1.00000x in dequantize-plus-matmul and 0.98532-1.02406x in backward.

Thus there is no material recurrent or user-visible gain to justify an SM103 launch rule.

### Correctness and boundary observations

- All 780 direct forced outputs (156 cells x five geometries) were finite and bit-exact to the baseline.
- The public allocating and `.out` paths were bit-exact to the direct baseline in all 156 cells.
- All forced outputs in the 8 dequantize-plus-matmul and 8 actual backward cells were finite and bit-exact to baseline.
- The baseline public empty-input probe fails with the existing zero-grid `invalid argument` at `csrc/ops.cu`; every measured geometry preserves that behavior. Fixing empty input is unrelated to the performance hypothesis and is intentionally not included here.

### Decision

The current 512-packed-byte/64-thread launch is already the best or tied configuration across the measured B300 region. Larger CTAs regress, and the only neutral larger-tile option has no direct, recurrent, or end-to-end benefit beyond noise. This investigation is therefore closed as a measurement NO-GO: no production edit, no branch commit, and no draft PR.
